From cc2a7969f2adb27ffbf21d58a0e9634faa522059 Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Mon, 8 May 2023 20:33:17 +0200
Subject: [PATCH 1005/1006] alpine: add bootcmd and runcmd

boocmd runs as early as possible, while runcmd runs late, during final.
---
 lib/tiny-cloud/cloud/alpine/init       |  3 ++-
 lib/tiny-cloud/user-data/alpine-config | 17 +++++++++++++
 tests/tiny-cloud-alpine.test           | 35 +++++++++++++++++++++++++-
 3 files changed, 53 insertions(+), 2 deletions(-)

diff --git a/lib/tiny-cloud/cloud/alpine/init b/lib/tiny-cloud/cloud/alpine/init
index dad1df7..087a6b8 100644
--- a/lib/tiny-cloud/cloud/alpine/init
+++ b/lib/tiny-cloud/cloud/alpine/init
@@ -31,8 +31,9 @@ INIT_ACTIONS_MAIN="$(insert_after save_userdata \
     load_userdata_handler \
     $INIT_ACTIONS_MAIN)"
 INIT_ACTIONS_MAIN="$(insert_after set_hostname \
-    "userdata_ntp userdata_apk_cache userdata_apk_repositories userdata_packages" \
+    "userdata_bootcmd userdata_ntp userdata_apk_cache userdata_apk_repositories userdata_packages" \
     $INIT_ACTIONS_MAIN)"
+INIT_ACTIONS_FINAL="$INIT_ACTIONS_FINAL userdata_runcmd"
 
 add_once() {
     local file="$1"
diff --git a/lib/tiny-cloud/user-data/alpine-config b/lib/tiny-cloud/user-data/alpine-config
index 1140366..0f1c908 100644
--- a/lib/tiny-cloud/user-data/alpine-config
+++ b/lib/tiny-cloud/user-data/alpine-config
@@ -1,6 +1,15 @@
 # Script UserData Functions
 # vim:set ts=4 et ft=sh:
 
+init__userdata_bootcmd() {
+    # run bootcmd
+    local bootcmds="$(imds user-data/bootcmd)"
+    for i in $bootcmds; do
+        local cmd="$(imds user-data/bootcmd/"$i")"
+        sh -c "$cmd"
+    done
+}
+
 init__userdata_ntp() {
     local ntp_enabled="$(imds user-data/ntp/enabled)"
     if [ "$ntp_enabled" != "yes" ] && [ "$ntp_enabled" != "true" ]; then
@@ -92,3 +101,11 @@ init__userdata_packages() {
         $MOCK apk add $pkgs
     fi
 }
+
+init__userdata_runcmd() {
+	local runcmds="$(imds user-data/runcmd)"
+	for i in $runcmds; do
+		local cmd="$(imds user-data/runcmd/$i)"
+		sh -c "$cmd"
+	done
+}
diff --git a/tests/tiny-cloud-alpine.test b/tests/tiny-cloud-alpine.test
index 78705a7..9078e62 100755
--- a/tests/tiny-cloud-alpine.test
+++ b/tests/tiny-cloud-alpine.test
@@ -9,6 +9,7 @@ lib="$srcdir"/lib/tiny-cloud/cloud/alpine/init
 
 init_tests \
 	set_network_config_network_interfaces \
+	userdata_bootcmd \
 	userdata_ntp \
 	userdata_ntp_busybox \
 	userdata_ntp_openntpd \
@@ -16,7 +17,8 @@ init_tests \
 	userdata_apk_repositories \
 	userdata_apk_repositories_version \
 	userdata_apk_repositories_version_auto_edge \
-	userdata_packages
+	userdata_packages \
+	userdata_runcmd
 
 
 set_network_config_network_interfaces_body() {
@@ -49,6 +51,19 @@ set_network_config_network_interfaces_body() {
 		cat etc/resolv.conf
 }
 
+userdata_bootcmd_body() {
+	fake_userdata_nocloud <<-EOF
+		#alpine-config
+		bootcmd:
+		  - echo foo
+		  - echo bar
+	EOF
+	atf_check \
+		-e match:"userdata_bootcmd .*DONE" \
+		-o match:"^foo$" -o match:"^bar$" \
+		tiny-cloud main
+}
+
 userdata_ntp_body() {
 	fake_userdata_nocloud <<-EOF
 		#alpine-config
@@ -174,3 +189,21 @@ userdata_packages_body() {
 		-o match:"apk add .*vim" \
 		tiny-cloud main
 }
+
+userdata_runcmd_body() {
+	fake_userdata_nocloud <<-EOF
+		#alpine-config
+		runcmd:
+		  - echo foo
+		  - echo bar
+	EOF
+	# run main phase to extract the user data
+	atf_check \
+		-e match:"save_userdata .*DONE" \
+		-o ignore \
+		tiny-cloud main
+	atf_check \
+		-e match:"userdata_runcmd .*DONE" \
+		-o match:"^foo$" -o match:"^bar$" \
+		tiny-cloud final
+}
-- 
2.40.1

